support-workers/src/main/scala/com/gu/zuora/subscriptionBuilders/DigitalSubscriptionBuilder.scala

```diff
@@ -5,90 +5,166 @@ import java.util.UUID
+  }
 
-    val productRatePlanId = validateRatePlan(digitalPack.productRatePlan(environment, digitalPack.readerType), digitalPack.describe)
+  def buildPurchase(
+    today: LocalDate,
+    productRatePlanId: ProductRatePlanId,
+    requestId: UUID,
+    readerType: ReaderType,
+    purchase: SubscriptionPurchase
+  )(implicit ec: ExecutionContext): BuildResult = {
+
+    val (contractAcceptanceDelay, autoRenew, initialTerm) = if (readerType == Gift)
+      (0, false, purchase.billingPeriod.monthsInPeriod)
+    else
+      (purchase.config.defaultFreeTrialPeriod + purchase.config.paymentGracePeriod, true, 12)
+
+    val contractAcceptanceDate = today.plusDays(contractAcceptanceDelay)
 
     val subscriptionData = buildProductSubscription(
       requestId,
       productRatePlanId,
       contractAcceptanceDate = contractAcceptanceDate,
-      contractEffectiveDate = contractEffectiveDate,
-      readerType = digitalPack.readerType
+      contractEffectiveDate = today,
+      readerType = readerType,
+      autoRenew = autoRenew,
+      initialTerm = initialTerm
     )
 
-    subscriptionPaymentType match {
-      case SubscriptionPaymentDirect(_, maybePromoCode, country, promotionService) =>
-        EitherT.fromEither[Future](
-          applyPromoCode(promotionService, maybePromoCode, country, productRatePlanId, subscriptionData)
-            .left.map(Left.apply)
-        )
-      case SubscriptionPaymentCorporate(redemptionData, getCodeStatus) =>
-        withRedemption(subscriptionData.subscription, redemptionData, getCodeStatus)
-          .map(subscription => subscriptionData.copy(subscription = subscription))
-          .leftMap(Right.apply)
+    val withRedemptionCode = generateRedemptionCode(readerType, purchase.billingPeriod, subscriptionData)
+    val withPromoApplied = applyPromoCode(purchase.promotionService, purchase.maybePromoCode, purchase.country, productRatePlanId, withRedemptionCode)
+
+    EitherT.fromEither[Future](withPromoApplied.left.map(Left.apply))
+  }
+
+  def generateRedemptionCode(readerType: ReaderType, billingPeriod: BillingPeriod, subscriptionData: SubscriptionData) = {
+    if (readerType == Gift) {
+      val code = GiftCodeGenerator.randomGiftCodes.next().withDuration(GiftDuration.fromBillingPeriod(billingPeriod))
+      subscriptionData.copy(
+        subscription = subscriptionData.subscription.copy(redemptionCode = Some(code.value))
+      )
     }
+    else
+      subscriptionData
+  }
 
+  def buildRedemption(
+    today: LocalDate,
+    productRatePlanId: ProductRatePlanId,
+    requestId: UUID,
+    readerType: ReaderType,
+    redemption: SubscriptionRedemption
+  )(implicit ec: ExecutionContext): BuildResult = {
+
+    readerType match {
+      case Corporate => buildCorporateRedemption(
+        today,
+        productRatePlanId,
+        requestId,
+        redemption.redemptionData.redemptionCode,
+        redemption.getCodeStatus
+      )
+      case Gift => buildGiftRedemption(
```

Comment:
I have some vague nervousness that we are using Gift in the state to both mean buy a gift and redeem a gift, and it's also a field on the subscription itself in zuora.  Everywhere before, readertype just went directly from the step function into zuora, but now it is conditional on a separate field (payment details) as to what it really means.
I think this is complicated by the fact that corporate redemption actually creates a zuora sub, wheras gift redemption just modifies a previously created one. 

Perhaps this just mostly inherent complexity due to the new offerings and the way we store in zuora, but even if we don't think the state model needs changing, we could still e.g. rename `Gift` to `GifterOrGiftee` or similar to make gotchas less likely in future.  Any thoughts?

Reply:
Maybe we should replace `Gift` with `Gifter` and `Giftee`? That might be an issue for downstream issues (primarily reporting) because we've been using `Gift` for GW gifters but it would be more accurate.

Reply:
Systems that could be affected:
- manage
- renewal functionality on subscribe
- DL reporting

Reply:
After discussion we are going to leave this for now